services/continuousquery/service.go

```diff
@@ -0,0 +1,288 @@
+/*
+Copyright 2022 Huawei Cloud Computing Technologies Co., Ltd.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package continuousquery
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/influxdata/influxdb/logger"
+	query2 "github.com/influxdata/influxdb/query"
+	"github.com/openGemini/openGemini/engine/index/tsi"
+	"github.com/openGemini/openGemini/lib/statisticsPusher/statistics"
+	"github.com/openGemini/openGemini/open_src/influx/influxql"
+	"github.com/openGemini/openGemini/open_src/influx/meta"
+	"github.com/openGemini/openGemini/open_src/influx/query"
+	"github.com/openGemini/openGemini/services"
+	"go.uber.org/zap"
+)
+
+const (
+	// DefaultReportTime is the default time interval for reporting stats.
+	DefaultReportTime     = 5 * time.Minute
+	DefaultInnerChunkSize = 1024
+)
+
+// ContinuousQuery is a struct for cq info, LastRun etc.
+type ContinuousQuery struct {
+	Database string
+	Info     *meta.ContinuousQueryInfo
+	HasRun   bool
+	LastRun  time.Time
+	q        *influxql.SelectStatement
+}
+
+// Service include query and store engine for CQ
+type Service struct {
+	services.Base
+
+	MetaClient interface {
+		Databases() map[string]*meta.DatabaseInfo
+	}
+
+	mu            sync.RWMutex
+	lastRuns      map[string]time.Time
+	QueryExecutor *query.Executor
+	RunInterval   time.Duration
+}
+
+// NewService creates a new Service instance named continuousQuery
+func NewService(interval time.Duration) *Service {
+	s := &Service{
+		lastRuns:      map[string]time.Time{},
+		RunInterval:   DefaultReportTime,
+		QueryExecutor: query.NewExecutor(),
+	}
+	s.Init("continuousQuery", interval, s.handle)
+	return s
+}
+
+func (s *Service) handle() {
+	logger, logEnd := log.NewOperation(s.Logger.GetZapLogger(), "continuous query check", "continuousQuery_check")
+	defer logEnd()
+	s.Logger.Info("Starting continuous query service")
+	if s.hasContinuousQuery() {
+		// get all databases
+		dbs := s.MetaClient.Databases()
+		// gets CQs from the meta and runs them.
+		if dbs == nil {
+			logger.Warn("No database found")
+		}
```

Comment:
```go 
logger, logEnd := log.NewOperation(s.Logger.GetZapLogger(), "continuous query check", "continuousQuery_check")
defer logEnd()
s.Logger.Info("Starting continuous query service")
if !s.hasContinuousQuery() {
	return
}

// get all databases
dbs := s.MetaClient.Databases()
// gets CQs from the meta and runs them.
if dbs == nil {
	return
}

for ...
......		
```

Reply:
It seems like something wrong with the following code in services/base.go :
```
func (s *Base) run() {
	if s.Interval == 0 {
		return
	}

	ticker := time.NewTicker(s.Interval)
	defer ticker.Stop()

	for {
		select {
		case <-s.done:
			return
		case <-ticker.C:
			if s.handle != nil {
				s.handle()
			}
		}
	}
}
```
I will check it tommorrow.

Reply:
@Benevor

Reply:
Everything is correct, when you remove the for loop in the `handle()`. Don't break the point in `run()`. Directly break the point in the `handle()` corresponding to the cq service. You will find that the `handle()` of cq will be called every second instead of every 30 minutes. So just remove the loop statement.